activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -101,7 +101,12 @@ def replace_keys(record)
         end
 
         def primary_key(klass)
-          reflection.association_primary_key(klass)
+          inverse = inverse_reflection_for(klass)
+          inverse && inverse.options[:primary_key] || reflection.association_primary_key(klass)
+        end
+
+        def has_many_reflection
```

Comment:
Can you explain why this method is here?

Reply:
Because `ActiveRecord::Associations::Association#set_inverse_instance` set wrong primary_key value into association foreign_key value when has_many association have `primary_key` option like below.
And it also happen belongs_to_association as belongs_to_polymophic_association.
But, this solution it isn't good because belongs_to have to se `inverse_of` option.

```rb
class Product < ActiveRecord::Base
  has_many :pictures, as: :imageable, primary_key: :code, inverse_of: :imageable
end

class Picture < ActiveRecord::Base
  belongs_to :imageable, polymorphic: true, inverse_of: :pictures
end


def set_inverse_instance(record)
  if inverse = inverse_association_for(record)
    inverse.inversed_from(owner)
  end
  record
end

def inversed_from(record)
  replace_keys(record)
  super
end

def replace_keys(record)
  owner[reflection.foreign_key] = record ? record._read_attribute(primary_key(record.class)) : nil
end

def primary_key(klass)
  # inverse.options[:primary_key] is :code
  # reflection.association_primary_key(klass) is :id
  inverse = inverse_reflection_for(klass)
  inverse && inverse.options[:primary_key] || reflection.association_primary_key(klass)
end
````

Reply:
`has_many_reflection` method is not need. It was added by wrong commit. I remove it.

Reply:
or other idea

```rb
module ActiveRecord
  module Associations
    # = Active Record Belongs To Association
    class BelongsToAssociation < SingularAssociation #:nodoc:
      ...

      def replace_keys(record)
        owner[reflection.foreign_key] = record ? record._read_attribute(inverse_primary_key(record) || primary_key(record.class)) : nil
      end

      # get primary key from inverse association
      def inverse_primary_key(record)
        inverse = inverse_reflection_for(record.class) || reverse_inverse_reflection_for(record)
        inverse && inverse.options[:primary_key]
      end

      # do not need inverse_of option on belong_to :xxxx, polymophic: true
      def reverse_inverse_reflection_for(record)
        record._reflections.values.find { |record_reflection| record_reflection.options[:inverse_of] == reflection.name }
      end

      def primary_key(klass)
        reflection.association_primary_key(klass)
      end
      
      ...
    end
  end
end
```